larpandora/LArPandoraEventBuilding/LArPandoraShower/Tools/ShowerESTAREnergy_tool.cc

```diff
@@ -0,0 +1,293 @@
+  {
+  }
+
+  int ShowerESTAREnergy::CalculateElement(const art::Ptr<recob::PFParticle>& pfparticle,
+        art::Event& Event,
+        reco::shower::ShowerElementHolder& ShowerEleHolder
+  ){
+
+    if(fVerbose){
+      std::cout << "~~~~~~~~~~~~~~~~~~~~~~~~~~~~ Shower ESTAR Energy Tool ~~~~~~~~~~~~~~~~~~~~~~~~~~~~" << std::endl;
+      if(fSCECorrectEField){std::cout << "Applying SCE correction" << std::endl;}
+    }
+      
+    // Get the associated pfParicle vertex PFParticles
+    auto const pfpHandle = Event.getValidHandle<std::vector<recob::PFParticle>>(fPFParticleLabel);
+
+    // Get the clusters
+    auto const clusHandle = Event.getValidHandle<std::vector<recob::Cluster>>(fPFParticleLabel);
+
+    const art::FindManyP<recob::Cluster>& fmc = ShowerEleHolder.GetFindManyP<recob::Cluster>(pfpHandle, Event, fPFParticleLabel);
+    std::vector<art::Ptr<recob::Cluster> > clusters = fmc.at(pfparticle.key());
+
+    // Get the hits association from the clusters
+    const art::FindManyP<recob::Hit>& fmhc = ShowerEleHolder.GetFindManyP<recob::Hit>(clusHandle, Event, fPFParticleLabel);
+    
+    std::map<geo::PlaneID::PlaneID_t, std::vector<art::Ptr<recob::Hit>>> planeHits;
+
+    // Get the spacepoints
+    auto const spHandle = Event.getValidHandle<std::vector<recob::SpacePoint>>(fPFParticleLabel);
+
+    const art::FindManyP<recob::SpacePoint>& fmsp = ShowerEleHolder.GetFindManyP<recob::SpacePoint>(pfpHandle, Event, fPFParticleLabel);
+    std::vector<art::Ptr<recob::SpacePoint> > spacepoints = fmsp.at(pfparticle.key());
+
+    // Get the hits association from the spacepoints
+    const art::FindManyP<recob::Hit>& fmhsp = ShowerEleHolder.GetFindManyP<recob::Hit>(spHandle, Event, fPFParticleLabel);
+
+    auto const clockData = art::ServiceHandle<detinfo::DetectorClocksService const>()->DataFor(Event);
+    auto const detProp   = art::ServiceHandle<detinfo::DetectorPropertiesService const>()->DataFor(Event, clockData);
+    
+    // Get E-field -  Electric Field in the drift region in KV/cm
+    nominal_Efield  = detProp.Efield(); // Nominal E-field 
+   
+    // Not all hits from the clusters will match to Space points. If that's the case let's use some sort of shower centre
+    // Get shower centre 
+    TVector3 showercentre = IShowerTool::GetLArPandoraShowerAlg().ShowerCentre(spacepoints);
+    
+    // Get the charge weighted shower centre - sum(charge*position)/sum(charge)
+    // Charge is obtained from the lifetime corrected hits and hits are obtained from the spacepoints 
+    TVector3 chargecentre = IShowerTool::GetLArPandoraShowerAlg().ShowerCentre(clockData, detProp, spacepoints, fmhsp);
+    
+    // Get space charge corrected E-field
+    if(fSCECorrectEField){
+      // Check the shower centre exists and is sensible
+      // Think the charge weighted centre can take nan values which causes the E-field calculation to fail
+      if(showercentre.Mag() >= 0 && chargecentre.Mag() >=0){ 
```

Comment:
I believe there is an STL function for checking nan: https://en.cppreference.com/w/cpp/numeric/math/isnan